Add Docker documentation (#5150)
Particular/docs.particular.net: servicecontrol/containerization/index.md

### Transports

Each supported transport and topology for ServiceControl has further been broken down into it's own set of images. The following transport and topologies are available on dockerhub:

* [Amazon SQS](https://hub.docker.com/search?q=servicecontrol.amazonsqs&type=image)
* [Azure Service Bus](https://hub.docker.com/search?q=servicecontrol.azureservicebus&type=image)
* [Azure Storage Queues](https://hub.docker.com/search?q=servicecontrol.azurestorageque&type=image)
* [RabbitMQ Conventional Topology](https://hub.docker.com/search?q=servicecontrol.rabbitmq.conven&type=image)
* [RabbitMQ Direct Topology](https://hub.docker.com/search?q=servicecontrol.rabbitmq.direct&type=image)
* [Sql Server](https://hub.docker.com/search?q=servicecontrol.sqlserver&type=image)

## Parameters

All parameters that the ServiceControl instance supports can be provided by passing the parameter as an environment variable. E.g.

> `-e "ServiceControl/Port=12345"`

Reviewer: ```suggestion
`-e "ServiceControl/Port=12345"`

The available parameters are

- ServiceControl/Port
- ServiceControl/ConnectionString
- ServiceControl.Audit/ConnectionString
- Monitoring/ConnectionString
- ServiceControl/LicenseText

The connectionstring is required by ServiceControl to connect to the transport.

These paramaters can also be added by using a standard Docker environment file. Every parameter has its own line and does not need to be enclosed by  quotes. It can then be used by Docker as follows:

`docker run --env-file servicecontrol.env [dockerimage]`

An example of the servicecontrol.env file:

```
# License text
ServiceControl/LicenseText=<?xml version="1.0" encoding="utf-8"?><license id="..."></license>

# Connection string
ServiceControl/ConnectionString=data source=server,1433; user id=sa; password=[password]; Initial Catalog=servicecontrol
```
```

---
title: Running ServiceControl in containers
reviewed: 2020-12-10
---

Docker images for ServiceControl exist on Dockerhub under the [Particular organization](https://hub.docker.com/u/particular). These can be used to run ServiceControl in docker containers. These docker images are only available for Windows due to ServiceControl's current dependency on Windows ESENT storage.

> NOTE: The current base Windows 2016 operating system images for ServiceControl do not support notifying the application that it is being shut down when running `docker stop`. This means ServiceControl cannot perform a graceful shutdown which in very few cases could result in manual intervention to repair the underlying datastore.

Reviewer: ```suggestion
NOTE: The base Windows operating system images for ServiceControl do not support notifying the application that it is being shut down when running `docker stop`. This means ServiceControl cannot perform a graceful shutdown which could result in a corrupt underlying datastore.
```

---
title: Running ServiceControl in containers
reviewed: 2020-12-10
---

> Docker images for ServiceControl exist on Dockerhub under the [Particular organization](https://hub.docker.com/u/particular). These can be used to run ServiceControl in docker containers. These docker images are only available for Windows.

Reviewer 2: ```suggestion
Docker images for ServiceControl exist on Dockerhub under the [Particular organization](https://hub.docker.com/u/particular). These can be used to run ServiceControl in docker containers. These docker images are only available for Windows due to ServiceControl its current dependency on Windows ESENT storage.
```

> NOTE: When shutting down a Docker container it will shut down all processes after 5 seconds. ServiceControl needs graceful shutdown or its embedded RavenDb database can get corrupted. Fixing this behavior is currently worked on and will be released as soon as possible.

Reviewer 2: ```suggestion
NOTE: The current base Windows 2016 operating system images for ServiceControl do not support notifying the application that it is being shut down when running `docker stop`. This means ServiceControl cannot perform a graceful shutdown which in very few cases could result in manual intervention to repair the underlying datastore.
```

## Containers overview

ServiceControl is split into multiple docker images for each instance type. These are:

* The error instance
* The audit instance
* The monitoring instance

> The images for each of these containers are further split into an init container and a runtime container. This means that there are 6 images for all instances of ServiceControl.

Reviewer 2: ```suggestion
The images for each of these containers are further split into an init container and a runtime container per transport.
```

* [Azure Service Bus](https://hub.docker.com/search?q=servicecontrol.azureservicebus&type=image)
* [Azure Storage Queues](https://hub.docker.com/search?q=servicecontrol.azurestorageque&type=image)
* [RabbitMQ Conventional Topology](https://hub.docker.com/search?q=servicecontrol.rabbitmq.conven&type=image)
* [RabbitMQ Direct Topology](https://hub.docker.com/search?q=servicecontrol.rabbitmq.direct&type=image)
* [Sql Server](https://hub.docker.com/search?q=servicecontrol.sqlserver&type=image)

## Parameters

All parameters that the ServiceControl instance supports can be provided by passing the parameter as an environment variable. E.g.

`-e "ServiceControl/Port=12345"`

The most common parameters used are likely to be:

* ServiceControl/ConnectionString
* ServiceControl/LicenseText
* ServiceControl.Audit/ConnectionString
* ServiceControl.Audit/LicenseText
* ServiceControl.Audit/ServiceControlQueueAddress
* ServiceControl/RemoteInstances
* Monitoring/ConnectionString
* Monitoring/LicenseText

NOTE: For hosting in Azure Container Instances parameter names are supported that use only underscores, e.g. ServiceControl_Audit_LicenseText

> These parameters can also be added by using a standard Docker environment file. Every parameter has its own line and does not need to be enclosed by  quotes. It can then be used by Docker as follows:

Reviewer 2: ```suggestion
These parameters can also be added by using a standard Docker environment file. Every parameter has its own line and does not need to be enclosed by quotes. It can then be used by Docker as follows:
```

`docker run --env-file servicecontrol.env [dockerimage]`

An example of the `servicecontrol.env` file:

```
# License text
ServiceControl/LicenseText=<?xml version="1.0" encoding="utf-8"?><license id="..."></license>

# Connection string
ServiceControl/ConnectionString=data source=[server],1433; user id=username; password=[password]; Initial Catalog=servicecontrol

# Remote audit instances

> ServiceControl/RemoteInstances=[{'api_uri':'http://[host_ip_address]:44444/api'}]

Reviewer 2: ```suggestion
ServiceControl/RemoteInstances=[{'api_uri':'http://[hostname]:44444/api'}]
```

Volume mappings can be used with ServiceControl for configuration files, storing logfiles and possibly reading a license file. More information on data storage can be found on the [Docker website](https://docs.docker.com/storage/).

### Configuration and Logging

Every image has a folder in `c:\data\` that can be mapped to a folder outside of the container. 

`docker run -v c:\servicecontrol\:c:\data\ [imagename]`

As a result, that container will write both logfiles and its database into that folder in respectively the `Logs` and `DB` folders. Another option is to map each folder specifically, for more control over their locations and divide them over separate drives, for example:

`docker run -v c:\servicecontroldatabase\:c:\data\DB\ -v d:\servicecontrollogfiles\:c:\data\logs\ [imagename]`

### License file

The license file can be provided via an environment variable on the command-line or via an environment file. Another options is to map the folder where ServiceControl will look for the license file.

> `docker run -v c:\servicecontrol\license\:%PROGRAMDATA%\ParticularSoftware\ [imagename]`

Reviewer 2: This doesn't seem right:

The ENVVAR would be applied BEFORE the volume mapping. The right part is the container location. That is fixed to `C:\ProgramData`, the left part is the host.

Isn't this mapping failing if the host would have Program Data on D: ?

This is the mapping I use in the sample:

    ${PROGRAMDATA}/ParticularSoftware:C:/ProgramData/ParticularSoftware

Author: The host could never have it on D as they can't change where the container internally stores the data. We may as well hardcode the exact path here because it is not possible for anyone to change the container's location.

> As the above command shows, ServiceControl will look for the license file in OS specific folders as can be found [in the documentation](/nservicebus/licensing/#license-management-machine-wide-license-location).

Reviewer 2: ```suggestion
As the above command shows, ServiceControl will look for the license file in OS specific folders as can be found [in the license documentation](/nservicebus/licensing/#license-management-machine-wide-license-location).
```

## ServiceControl maintenance

If maintenance is required on the embedded RavenDb database of ServiceControl, this can be enabled on Docker by

1. Stopping the ServiceControl container
2. Starting a new container by adding the `--maintenance --portable` parameters and adding an additional port mapping for port 33334.

> `docker run -it -d particular/servicecontrol.sqlserver.init-windows:4 --maintenance --portable`

Reviewer 2: ```suggestion
`docker run --interactive --tty --detach particular/servicecontrol.sqlserver.init-windows:4 --maintenance --portable`
```

* [Amazon SQS](https://hub.docker.com/search?q=servicecontrol.amazonsqs&type=image)
* [Azure Service Bus](https://hub.docker.com/search?q=servicecontrol.azureservicebus&type=image)
* [Azure Storage Queues](https://hub.docker.com/search?q=servicecontrol.azurestorageque&type=image)
* [RabbitMQ Conventional Topology](https://hub.docker.com/search?q=servicecontrol.rabbitmq.conven&type=image)
* [RabbitMQ Direct Topology](https://hub.docker.com/search?q=servicecontrol.rabbitmq.direct&type=image)
* [Sql Server](https://hub.docker.com/search?q=servicecontrol.sqlserver&type=image)

## Running ServiceControl using Docker

This section uses the SqlServer transport as an example on how to run ServiceControl using docker. The same steps are applicable to other transports.

Each instance type of ServiceControl has a distinct init and runtime container.

* The error instance of ServiceControl can be run as stand-alone
* The audit instance needs an error instance to be running too
* The monitoring instance can be run as stand-alone

### Error instances

The first step is to create the required infrastructure using the `init` container:

> `docker run -e "ServiceControl/ConnectionString=[connectionstring]" -e 'ServiceControl/LicenseText=[licensecontents]' -v c:/data/:c:/data/ -v c:/logs/:c:/logs/ -d particular/servicecontrol.sqlserver.init-windows:4`

Reviewer: It's not entirely clear this connectionstring var is also used with Azure Service Bus. Maybe mention that explicitely. And explain all available vars.

Author: Added a section explaining that all variables are environment variables
